fw/http.c

```diff
@@ -1320,6 +1478,112 @@ tfw_http_send_resp(TfwHttpReq *req, int status, const char *reason)
 		tfw_h1_send_resp(req, status);
 }
 
+static void
+tfw_http_send_resp2(TfwHttpReq *req, TfwStr *msg, int status)
+{
+	if (TFW_MSG_H2(req)) {
+		tfw_h2_send_resp2(req, msg, status, 0);
+	} else {
+		TfwCliConn *cli_conn = (TfwCliConn *)req->conn;
+
+		WARN_ONCE(!list_empty_careful(&req->msg.seq_list),
+			  "Request is already in seq_queue\n");
+		tfw_stream_unlink_msg(req->stream);
+		spin_lock(&cli_conn->seq_qlock);
+		list_add_tail(&req->msg.seq_list, &cli_conn->seq_queue);
+		spin_unlock(&cli_conn->seq_qlock);
+
+		tfw_h1_send_resp2(req, msg, status);
+	}
+}
+
+static void
+tfw_http_req_redir(TfwHttpReq *req, int status, const char *pattern)
+{
+	size_t url_len = strlen(pattern) + 1;
+	char *url = kmalloc(url_len, GFP_KERNEL);
+	char *date_val = *this_cpu_ptr(&g_buf);
+	char *url_p = url;
+	const char *pattern_end = pattern + strlen(pattern);
+	TfwStr *c, *end;
+	char *status_line;
+
+	tfw_http_prep_date(date_val);
+
+#define TFW_STRCPY(from)                                                 \
+do {                                                                     \
+	size_t dif = url_p - url;                                            \
+	url = krealloc(url, url_len + (from)->len, GFP_KERNEL);              \
+	if (!url) {                                                          \
+		T_WARN("HTTP: unable to allocate memory for redirection url\n"); \
+		return;                                                          \
+	}                                                                    \
+	url_p = url + dif;                                                   \
+	TFW_STR_FOR_EACH_CHUNK(c, (from), end) {                             \
+		memcpy_fast(url_p, c->data, c->len);                             \
+		url_p += c->len;                                                 \
+	}                                                                    \
+} while (0)
+
+	for (; *pattern; ++pattern) {
+		if (*pattern == '$') {
```

Comment:
This is not good to do all this stuff in runtime, especially in softirq and especially with `GFP_KERNEL`. Let's introduce helping flags to `TfwHttpRedir`: which variables are used in the rule (say up to 8) and an array of variable positions and types (the array is of variable length, which isn't a problem on config type). `TfwHttpRedir->url` having variables should be `TfwStr`.

We need to extend `tfw_http_req_t` with the array of variable (see https://github.com/tempesta-tech/tempesta/issues/907). At some point we might arrange all the fields of the the request, which can be referenced by variables, in a special headers array (this is quite intuitive for HTTP/2), but now just let's keep references to TfwStr and please write a comment that it's good to optimize to avoid indirect references.

Variables in a compiled pattern for `TfwHttpRedir` should be just static indexes in the variables array in `tfw_http_req_t`.

In this function we just need to check the flags and merge TfwStr by the positions where the variables are presented.